File: server-side-pong/player_paddle.py
from lib import Point, Rect, is_colliding_ball_paddle
from ball import Ball


class PlayerPaddle:
    shape: Rect
    y_vector: float

    def __init__(
            self, y_vector: float, shape: Rect):
        self.y_vector = y_vector
        self.shape = shape

    def move_up(self, ball: Ball):
        new_y = self.shape.y - self.y_vector
        new_pos = Point(self.shape.x, new_y)
        ball_center = Point(
            ball.shape.x + ball.radius_px, ball.shape.y + ball.radius_px)
        if is_colliding_ball_paddle(ball_center, ball.radius_px, new_pos,
           self.shape.width, self.shape.height) is not None:
            return
        if new_y < 0:
            new_y = 0
        self.shape.y = new_y

    def move_down(self, ball: Ball, arena_height: int):
        new_y = self.shape.y + self.y_vector
        new_pos = Point(self.shape.x, new_y)
        ball_center = Point(
            ball.shape.x + ball.radius_px, ball.shape.y + ball.radius_px)
        if is_colliding_ball_paddle(ball_center, ball.radius_px, new_pos,
           self.shape.width, self.shape.height) is not None:
            return
        if (new_y + self.shape.height) > arena_height:
            new_y = arena_height - self.shape.height
        self.shape.y = new_y

File: frontend/src/static/playerPaddle.ts
import { arenaHeight, isCollidingBallPaddle, Point } from "./lib"
import { type Ball } from "./ball"

export class PlayerPaddle
{
    x: number
    y: number
    width: number
    height: number
    yVector: number
    color: string
    upPressed = false
    downPressed = false

    constructor(x: number, y: number, width: number, height: number, yVector: number, color: string) {
        this.x = x
        this.y = y
        this.width = width
        this.height = height
        this.yVector = yVector
        this.color = color
    }

    moveUp(ball: Ball) {
        let newY = this.y - this.yVector

        if (newY < 0) {
            newY = 0
        }
        const ballCenterX = ball.x + ball.radius
        const ballCenterY = ball.y + ball.radius
        if (isCollidingBallPaddle(ballCenterX, ballCenterY, ball.radius, this.x, newY,
            this.width, this.height) !== null) {
            return
        }
        this.y = newY
    }

    moveDown(ball: Ball) {
        let newY = this.y + this.yVector
        if (newY + this.height > arenaHeight) {
            newY = arenaHeight - this.height
        }
        const ballCenterX = ball.x + ball.radius
        const ballCenterY = ball.y + ball.radius
        if (isCollidingBallPaddle(ballCenterX, ballCenterY, ball.radius, this.x, newY,
            this.width, this.height) !== null) {
            return
        }
        this.y = newY
    }

    update(ball: Ball) {
        if (this.downPressed) {
            this.moveDown(ball)
        }
        if (this.upPressed) {
            this.moveUp(ball)
        }
    }

    draw(canvasCtx: CanvasRenderingContext2D) {
        canvasCtx.fillStyle = this.color
        canvasCtx.fillRect(this.x, this.y, this.width, this.height)
    }
}

export namespace player_paddle
{
    export enum hitSide
    {
        Top,
        Bottom,
        Left,
        Right,
        None
    } 
}
